TriblerGUI/widgets/trustpage/js/data_processor.js

```diff
@@ -0,0 +1,344 @@
+ * result of an array of reducers.
+ *
+ * @param {GraphResponseData} response - The response
+ * @param {Reducer[]} reducers
+ * @returns {Object} - The resulting object
+ */
+function reduceResponse(response, reducers) {
+
+    return reducers.reduce(function (interim, reducer) {
+        var additions = reducer(response, interim);
+
+        return Object.assign(interim, additions);
+    }, {});
+}
+
+/**
+ * Map the GraphResponseNode to GraphNode objects
+ *
+ * @param {GraphResponseData} response - The response
+ * @returns {{edges: GraphEdge[]}}
+ */
+function mapNodes(response) {
+    var nodes = response.nodes.map(function (node) {
+        return {
+            public_key: node.public_key,
+            total_up: node.total_up,
+            total_down: node.total_down,
+            page_rank: node.page_rank
+        }
+    });
+
+    return {
+        nodes: nodes
+    }
+}
+
+/**
+ * Map the directed GraphResponseEdge to GraphEdge objects
+ *
+ * @param {GraphResponseData} response - The response
+ * @param {Object} interim - The interim result (expects interim.nodes)
+ * @returns {{edges: GraphEdge[]}}
+ */
+function mapEdges(response, interim) {
+
+    var edges = response.edges.map(function (edge) {
+        return {
+            source: find(interim.nodes, "public_key", edge.from),
+            source_pk: edge.from,
+            target: find(interim.nodes, "public_key", edge.to),
+            target_pk: edge.to,
+            amount: edge.amount
+        }
+    });
+
+    return {
+        edges: edges
+    }
+}
+
+/**
+ * Combine the directed edges between nodes to one link per pair.
+ *
+ * @param {GraphResponseData} response - The response
+ * @param {Object} interim - The interim result (expects interim.nodes, interim.edges)
+ * @returns {{links: GraphLink[]}}
+ */
+function combineLinks(response, interim) {
+
+    var groupedEdges = groupBy(interim.edges, "source_pk");
+
+    // Make the combined links for all nodes
+    var combinedLinks = flatMap(interim.nodes, function (sourceNode) {
+        return (groupedEdges[sourceNode.public_key] || []).map(function (edge) {
+            var edgesFromTarget = groupedEdges[edge.target_pk];
+            var inverseEdge = edgesFromTarget !== undefined ? find(edgesFromTarget, "target_pk", sourceNode.public_key) : null;
+            var up = edge.amount,
+                down = inverseEdge ? inverseEdge.amount : 0,
+                empty = up + down === 0;
+
+            // Remove the inverse edge
+            if (inverseEdge) {
+                groupedEdges[edge.target_pk].splice(groupedEdges[edge.target_pk].indexOf(inverseEdge), 1);
+            }
+
+            return {
+                source_pk: edge.source_pk,
+                target_pk: edge.target_pk,
+                source: edge.source,
+                target: edge.target,
+                amount_up: up,
+                amount_down: down,
+                ratio: empty ? 1 : up / (up + down),
```

Comment:
You could place the fallback cases at the end by negating the condition, like you did above.